Solarkraft blog post 4: Hybrid Blockchain Monitors (#10)
konnov/protocols-made-fun: _posts/2024-06-24-solarkraft-part4.md

_Solarkraft has been developed in collaboration by [Igor Konnov][], [Jure Kukovec][], [Andrey Kuprianov][] and [Thomas Pani][]._

While the previous posts explain the current state of the project, in this one we take one step further, and explore the directions in which we plan to evolve blockchain runtime monitoring with Solarkraft. Throughout the post we are using the same [`timelock` contract][timelock] from `soroban-examples` that was used in [Part 2: "Guardians of the Blockchain"][part2]; please explore at least this post first to acquire the necessary context.

## Blockchain Runtime Monitors

> Runtime monitoring, also known as [Runtime verification][], is a well-established field, where many practical approaches have been devised and applied successfully. Based on this heritage, we proposed in [Part 2: "Guardians of the Blockchain"][part2] the first version Web3 runtime monitoring system for the Stellar blockchain, which is based on the [TLA+][] specification language, a well-established formalism for specifying and verifying distributed systems. 

Author: Thanks, done

Taking a step back from the concrete solution, let's try to answer the more abstract question: _What do we want to achieve with runtime monitors in blockchains?_ As eventually runtime monitors are going to be deployed and executed _on the live blockchain_, and they should satisfy the three requirements:

> - **Prevent safety violations** (_safety_): bad things, such as your token balance being stolen, should not happen. This is the primary goal of runtime monitors: react preventively, and abort unwanted executions.

Author: I tell in the beginning of the post that I explore the future directions, not the current state:) 

> While the previous posts explain the current state of the project, in this one we take one step further, and explore the directions in which we plan to evolve blockchain runtime monitoring with Solarkraft.

- **Detect liveness violations** (_liveness_): good things should be able to happen! E.g. you, as an account owner, should be able to withdraw your own balance. If a legitimate transaction keeps reverting, that's also a bug, not less severe than someone stealing your tokens.

> - **Detect unexpected behaviors** (_completeness_): same as code, specs are not perfect. If a spec author overlooked some behaviors, and they manifest themselves on the live blockchain, this may mean anything: from simple spec incompleteness, to an exploit being executed. Monitors should be configurable to either issue a simple warning, or to prevent such behaviors altogether.

Author: sorry, the semantics is different.

- mine: there are some overlooked behaviors, and there is a separate fact of those behaviors _unexpectedly_ showing up on the live blockchain, what to do?
- yours is weaker: there are some overlooked behaviors which manifest themselves on the live blockchain. 

I don't really like to merge these two into one... This will stay as is.

The problem we've observed with the previously employed specification approaches is that the specs of _what_ the method should do can easily be much larger than the actual implementation; compare e.g. this [ERC-721 Certora spec][] with the [ERC-721 implementation][] (don't forget to exclude comments when comparing). So we would like to add to the above the following soft requirement:

> - **Specify behaviors compactly and independently** (_compactness and modularity_): it is usually the case that a smart contract encompasses a lot of various aspects (e.g. authentication, authorization, storage management, math computations), as well is written/employed/reasoned about by various roles (e.g. smart contract developer, mathematician, architect, UI developer). All of those roles should be able to specify various aspects of the smart contract behavior as easily and as independently as possible.

Author: thanks:)

> So monitors should be able to specify both _safety_ and _liveness_ properties, be _complete_ wrt. the current and future system behaviors, and, preferably, also be _compact and modular_. For that we propose to apply a certain structure to the _direct monitors_ (those reasoning from cause to effect), as well as complement it with _reverse monitors_ (those going from effect to cause), thus combining them together in what we call _hybrid monitors_.

Reviewer: ```suggestion
So monitors should be able to specify both _safety_ and _liveness_ properties, be _complete_ wrt. the current and future system behaviors, and, preferably, also be _compact and modular_. 
For that we propose a conceptual separation of monitors into _direct monitors_ (those reasoning from cause to effect), and _reverse monitors_ (those going from effect to cause).
We can combine the two together in what we call _hybrid monitors_.
```

Author: A good suggestion, thanks!

## Direct Monitors

> Here we reason from the cause (method invocation) to the effect, but apply a structure which closely mimics in formal semantics what we expect to see when we program smart contracts. The essence of the structure is in the picture below:

Reviewer: ```suggestion
Here we reason from the cause (method invocation) to the effect, but apply a structure which closely mimics, in formal semantics, what we expect to see when we program smart contracts. The essence of the structure is in the picture below:
```

![Direct monitor specs](/img/DirectMonitors.png)

> 

Reviewer: ```suggestion
In other words, we distinguish three kinds of properties:
```

Author: makes sense; but I will make it "three kinds of conditions"; "properties" are too general for that...

> - `MustFail_i` is a condition under which the method is expected to fail. If _any_ of those conditions hold, the monitor fires, and checks that the method does indeed fail;

Reviewer: I'd personally avoid the "monitor _fires_" terminology in general

Author: Thanks! I have replaced "fired/fires" with "holds" in most cases.

- `MustPass_i` is a condition, under which the method is expected to succeed, _provided that_ none of the `MustFail_i` conditions fired. Each `MustPass_i` condition represents a separate happy path in the method invocation;

> - `MustHold_i` is a condition that should hold over past and next state variables, if the method invocation is successful (e.g. the tokens should be transferred). _All_ of `MustHold_i` should hold if the method is executed successfully.

Reviewer: ```suggestion
- `MustHold_i` is a condition that should hold after the method invocation is successful (e.g. the tokens should be transferred). Unlike the previous two categories, which reason only about the state of the system before the method invocation, these properties may reference both the post-method state, and the pre-method state._All_ of `MustHold_i` should hold if the method is executed successfully.
```

Author: Thanks, Jure!

```
Must<Fail|Pass|Hold>_<Method>_<ConditionDescription>
```

All predicates which refer to the same `<Method>` will be grouped, to create together one _method monitor_. Interpreted formally, the monitor should do the following when `<Method>` is invoked:

1. If any of `MustFail_i` conditions fire, check that method invocation reverts (otherwise, issue a warning / revert if configured to do so)
2. If none of `MustFail_i` conditions fired, but method invocation reverted, issue a warning (incomplete spec)
3. If none of  `MustFail_i` fired, and one of `MustPass_i` conditions fired, check that method invocation succeeds (otherwise, issue a warning)
3. If none of  `MustFail_i` fired, and none of `MustPass_i` conditions fired, but method invocation succeeded, issue a warning of an incomplete spec (or revert if configured to do so)
4. If method invocation succeeds, check that all of `MustHold_i` conditions hold on the pre- and post-states of the method invocation (otherwise, issue a warning / revert if configured to do so)

> Notice that above we apply _or_ as default connector for preconditions (`MustFail_i` / `MustPass_i`), and we apply _and_ as default connector for effects (`MustHold_i`). Thus, you may split preconditions/effects into separate predicates at your convenience, thus avoiding complicated logical structure inside predicates themselves.

Reviewer: ```suggestion
Notice that above we apply _or_ as default connector for preconditions (`MustFail_i` / `MustPass_i`), and we apply _and_ as default connector for effects (`MustHold_i`). Thus, you may split preconditions/effects into separate predicates at your convenience, avoiding complicated logical structure inside predicates themselves.
```

### Direct monitors for the Timelock contract

Having outlined the general structure of direct monitors, let's apply it to the [Timelock contract][timelock]. Direct monitors for Timelock's `deposit` and `claim` methods can be found in [deposit.tla][deposit] and [claim.tla][claim] respectively; below we depict only the structure of these monitor specifications (we omit `Must<Fail|Pass|Hold>` as well as the method names for clarity).

![Timelock direct monitors](/img/TimelockDirectMonitors.png)

As can be seen, a direct method monitor is decomposed into a collection of independent and small monitors, i.e. we did achieve our (soft) goal of _compactness and modularity_. Safety and liveness goals also seem to be satisfied:

> - _Safety_: Timelock's direct monitors guarantee the whole bunch of safety properties, a safety property is usually ensured by either <font style="background: #fd4d4d;">MustFail</font>, or <font style="background: #ffe342;">MustHold</font>, or a combination of both conditions. For example:

Reviewer: ```suggestion
- _Safety_: Timelock's direct monitors guarantee numerous safety properties. A safety property is usually ensured by either <font style="background: #fd4d4d;">MustFail</font>, or <font style="background: #ffe342;">MustHold</font>, or a combination of both conditions. For example:
```

Author: this is better, thanks!

  - The property "the Timelock contract receives the deposited funds from the claimant" is ensured by the combination of <font style="background: #fd4d4d;">NotEnoughBalance</font> and  <font style="background: #ffe342;">TokenTransferred</font> sub-monitors.
 - _Liveness_: Timelock's liveness properties are guaranteed by the <font style="background: #54c45e;">MustPass</font> conditions: 
   - Implicit in case of `deposit` (whatever doesn't fail, should succeed);
   - Explicit in case of `claim`: a claim happening before the time bound, when its kind is `Before`, should succeed due to <font style="background: #54c45e;">BeforeTimeBound</font> (provided all other conditions are met); similarly, a claim happening after the time bound, when its kind is `After`, should succeed due to <font style="background: #54c45e;">AfterTimeBound</font>.

Can the described approach of direct monitors be considered satisfactory? Please stop to think about it for a sec, before opening our answer below.

<details>
<summary> <b>Are direct monitors enough?</b></summary>

> <p>You may have guessed the answer: we believe NO! And here is an (incomplete) list of whys:</p>

Reviewer: ```suggestion
<p>You may have guessed the answer: we believe NO! And here is an (incomplete) list of reasons why:</p>
```

<ol>

> <li> A method may have a side effect, which was overlooked by the spec author. E.g. a boolean flag is set, which in another method allows to direct funds to another account.</li>

Reviewer: ```suggestion
<li> A method may have a side effect, which was overlooked by the spec author. E.g. a boolean flag gets set, which allows the redirection of funds to another account in a different method.</li>
```

Author: yeah, this is a better formulation!

> <li> Code evolves, but the spec stays as is; as a result a side effect above is introduced unintentionally, with the stale spec not accounting for it.</li>

Reviewer: ```suggestion
<li> Code evolves, but the spec stays as is; as a result a side effect, like the one above, is introduced unintentionally, with the stale spec not accounting for it.</li>
```

> <li> Internal state component is modified in multiple methods, in different ways. The specification about how the component should be updated is scattered in multiple places, and loopholes may easily creep in.</li>

Reviewer: ```suggestion
<li> The same internal state component is modified by multiple methods, in different ways. The specification of how the component should be updated is scattered across multiple places, and loopholes may easily creep in.</li>
```

Author: also good!

![Reverse monitor specs](/img/ReverseMonitors.png)

- `MonitorTrigger_i` is a condition which triggers (activates) the monitor. If _any_ of those conditions hold, the monitor is activated;
- `MonitorEffect_i` is a condition over past and next state variables, which specifies the effect that the the monitor, if activated, should check. _All_ of `MonitorEffect_i` should hold if the transaction is successful.

Similar to direct reasoning, in the above, `Monitor<Trigger|Effect>` is a prefix, which tells the monitor system how to interpret this predicate. The complete pattern for predicate names with these prefixes is as follows:

```
Monitor<Trigger|Effect>_<Monitor>_<ConditionDescription>
```

All predicates which refer to the same `<Monitor>` will be grouped, to create together one _effect monitor_. Interpreted formally, the monitor should do the following when activated:

> - If _any_ of `MonitorTrigger_i` conditions fire, check that _all_ `MonitorEffect_i` hold over the past and next states (otherwise, issue a warning / revert if configured to do so)

Reviewer: I don't remember if I made this comment on your original writeup, but how exactly is this different from `MustPass_ & MustHold_`? Even if it is, you really should describe the difference explicitly.

Author: I have added more explanations; thanks for pointing this out!

Taking a step back from the concrete solution, let's try to answer the more abstract question: _What do we want to achieve with runtime monitors in blockchains?_ As runtime monitors are eventually going to be deployed and executed _on the live blockchain_, and they should satisfy the following requirements:

- **Prevent safety violations** (_safety_): bad things, such as your token balance being stolen, should not happen. This is the primary goal of runtime monitors: react preventively, and abort unwanted executions.
- **Detect liveness violations** (_liveness_): good things should be able to happen! E.g. you, as an account owner, should be able to withdraw your own balance. If a legitimate transaction keeps reverting, that's also a bug, not less severe than someone stealing your tokens.
- **Detect unexpected behaviors** (_completeness_): same as code, specs are not perfect. If a spec author overlooked some behaviors, and they manifest themselves on the live blockchain, this may mean anything: from simple spec incompleteness, to an exploit being executed. Monitors should be configurable to either issue a simple warning, or to prevent such behaviors altogether.

> The problem we've observed with the previously employed specification approaches is that the specs of _what_ the method should do can easily be much larger than the actual implementation; compare e.g. this [ERC-721 Certora spec][] with the [ERC-721 implementation][] (don't forget to exclude comments when comparing). So we would like to add to the above the following soft requirement:

Reviewer 2: Do we have to compare with Certora? In general, I would like to avoid direct comparisons like that, since things in industry are moving very fast

  - The property "only the approved claimant may claim the deposit" is ensured by the <font style="background: #fd4d4d;">NonClaimant</font> sub-monitor;
  - The property "the Timelock contract receives the deposited funds from the claimant" is ensured by the combination of <font style="background: #fd4d4d;">NotEnoughBalance</font> and  <font style="background: #ffe342;">TokenTransferred</font> sub-monitors.
- _Liveness_: Timelock's liveness properties are guaranteed by the <font style="background: #54c45e;">MustPass</font> conditions: 
  - Implicit in case of `deposit` (whatever doesn't fail, should succeed);
  - Explicit in case of `claim`: a claim happening before the time bound, when its kind is `Before`, should succeed due to <font style="background: #54c45e;">BeforeTimeBound</font> (provided all other conditions are met); similarly, a claim happening after the time bound, when its kind is `After`, should succeed due to <font style="background: #54c45e;">AfterTimeBound</font>.

Can the described approach of direct monitors be considered satisfactory? Please stop to think about it for a sec, before opening our answer below.

<details>

> <summary> <b>Are direct monitors enough?</b></summary>

Reviewer 2: ```suggestion
<summary> <b>Are direct monitors sufficient?</b></summary>
```
